compiler/rustc_type_ir/src/opaque_ty.rs

```diff
@@ -0,0 +1,51 @@
+use rustc_macros::{HashStable_NoContext, TyDecodable, TyEncodable};
+use rustc_type_ir_macros::{TypeFoldable_Generic, TypeVisitable_Generic};
+
+use crate::inherent::*;
+use crate::{self as ty, Interner};
+
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = ""),
+    Hash(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = ""),
+    Debug(bound = ""),
+    Copy(bound = "")
+)]
+#[derive(TypeVisitable_Generic, TypeFoldable_Generic)]
+#[cfg_attr(feature = "nightly", derive(TyEncodable, TyDecodable, HashStable_NoContext))]
+pub struct OpaqueTypeKey<I: Interner> {
+    pub def_id: I::LocalDefId,
+    pub args: I::GenericArgs,
+}
+
+impl<I: Interner> OpaqueTypeKey<I> {
+    pub fn iter_captured_args(self, tcx: I) -> impl Iterator<Item = (usize, I::GenericArg)> {
```

Comment:
That `+ IntoIterator<Item: Deref<Target = ty::Variance>>` bound is required for `VariancesOf` because we can't return an RPIT that borrows from a local, which it does if we were to call `slice::iter()`.

I think I'd like to eventually make a trait that's like `IntoIterator` but which consolidates `ty::List` and `&[_]` with a single `into_iter()` function, and always returns the copied version of the data. Like something to turn `&[Ty]` into an iterator of `Ty` not `Ty`.